### ISSUE TITLE
Track.to_jams() implementations are no longer deterministic

### ISSUE THREAD

**Issue by lostanlen** [mir-dataset-loaders/mirdata#237]
Consider the following script:
```python
from mirdata import orchset
orchset.download()
track = orchset.Track('Beethoven-S3-I-ex1')

jam1 = track.to_jams()
custom_function()
jam2 = track.to_jams()
```

Independently of `custom_function`, it is safe to assume that:
1. the script should run without error
2. the value of `jam1` and `jam2` will be the same if we re-run the script
3. `jam1==jam2` is `True`

And, as of mirdata v0.2-beta, that is indeed the case.

However, PRs #212 and #230 break properties (1), (2) and (3). The issue is not limited to ORCHSET, but arises in 11 dataset modules:
* `beatles`
* `dali`
* `ikala`
* `medley_solos_db`
* `medleydb_melody`
* `medleydb_pitch`
* `orchset`
* `rwc_classical`
* `rwc_jazz`
* `rwc_popular`
* `salami`
* `tinysol`

Crucially, the bug arises despite the fact that `custom_function()` does not
modify any property of `track`. Depending on how vicious the `custom_function()`
is, `jam1==jam2` may be systematically `False`, or it may be `True` with some
nonzero probability (Heisenbug). It may be `True` on some machines but `False`
in others. 

In fact, as soon as the `data_home` folder (here, implicitly passed to `Track`
via the default kwarg `data_home=None`) is writable, the result of `jam1==jam2`
is nondeterministic; i.e., we cannot predict it before `jam2` is computed. Worse,
the `jam1` assignment may pass while the `jam2` assignment may fail.

The reason behind this bug is that the new implementation of `to_jams` for these
11 modules makes an unsafe call to either `librosa.get_duration` or`librosa.load`.
For example, here is the current implementation of `orchset.Track.to_jams()`:

```python
def to_jams(self):
    """Jams: the track's data in jams format"""
    metadata = {k: v for k, v in self._track_metadata.items() if v is not None}
    metadata['duration'] = librosa.get_duration(
        self.audio_mono[0], self.audio_mono[1]
    )
    return jams_utils.jams_converter(
        f0_data=[(self.melody, 'annotated melody')], metadata=metadata
    )
```

The call to `get_duration` is unsafe in the sense that it does not verify the
md5 checksum of the audio during loading. In fact, it does not even check
if the file exists and if the user has read permission. So `to_jams` may return
the wrong value if the file has been modified since `Track.__init__`.
Furthermore, `to_jams` may now raise the following exceptions: `FileNotFoundError`,
`PermissionError`, and `IOError`.

The only modules in which this bug does not arise are: `groove_midi` (because duration is a metadata field) and `gtzan_genre` (because duration is considered constant).
`guitarset` is also affected, but in a different way: i.e. via the JAMS file of the Track rather than the WAV files.

Such behavior is inconsistent across modules, not backwards
compatible, and undocumented.

See this gist for a plausible use case in which this bug would arise:
https://gist.github.com/lostanlen/ea38c0591bcc8406f15fd84f0bfdb84f

**Comment by lostanlen** [comment]
Here's what i suggest:

(1) in the short term (v0.2), stop prioritizing JAMS compliance for. Remove the call sites to `get_duration`. release v0.2 with the same behavior as v0.2-beta. I don't think it's a good idea to introduce so much inconsistent and undocumented behavior in the prospect of a release. In my opinion, `to_jams().validate()` can wait till v0.3

(2) in the midterm (v0.3?), write a `load_librosa` method in the base `Track` class, which does all the necessary checks against the index: `FileNotFoundError` (file has been deleted), `PermissionError` (we no longer have write access), and `OSError` (checksum is wrong). Likewise, we could implement `load_pretty_midi` and `load_jams` to protect `pretty_midi.load` and `jams.load` from returning nondeterministic values. Such protection could happen via a `@validated` decorator.

(3) in the long term (v0.4?), if we really want to pass duration to `to_jams()`, these protected loaders could be used to estimate duration in a way that is deterministic. (even though the `audio_*`, `midi`, and `jams` property are not `@cached_property`). It is also important to be able to set duration as read-only in `groove_midi`. At that stage, `to_jams()` will be deterministic and JAMS-compliant.

TL;DR: at the end of the day, i don't care that much if it's `to_jams()` or `to_jams(duration)` or `to_jams(duration=None)` or `to_jams(duration=0)` or whatever. But please, please, let's keep `to_jams` deterministic. Let's not sacrifice predictability of `to_jams().file_metadata.duration` for the sake of having something that passes `jams.JAMS.validate()`. It's better to have a return value that isn't fully initialized than to have a nondeterministic return value.

As of v0.2-beta, our `to_jams()` function was non-constant but deterministic. Now, it is non-constant and nondeterministic. I wish we reverted that change. The really long-term goal  (v1.0?) would be to have a constant return of `to_jams()`, but we're nowhere near there yet.

just my 2¢
i'll let @rabitt and @magdalenafuentes be the judge

**Comment by rabitt** [comment]
This was introduced in #212 

I"m going to make the call that, since I think we (so far) have very few jams users, that this is a "won't fix", and accept that the behavior changed silently between versions. So far no one complained about jams files not validating, and I expect that if someone had been manually adding duration to jams files created by mirdata, they would have opened an issue.
